### DISCUSSION TITLE
How to get a single custom sized thumbnail binary?

### DISCUSSION THREAD

**Discussion by Ghostbird** [microsoftgraph/msgraph-sdk-dotnet#2686] [Q&A] [answered]
I'm migrating from SDK 4 to SDK 5 and ran into the issue that I cannot find how to convert this code:

```csharp
      return await myDriveItemRequestBuilder
        .Thumbnails["0"]["c256x256"]
        .Content
        .Request()
        .GetAsync(cancellationToken);
```
Which generates a URI like:
> /me/drive/items/{item-id}/thumbnails/0/c256x256/content

I've managed to get to 

```csharp
      return await myDriveItemRequestBuilder
        .Thumbnails["0"]
        .GetAsync(request => request.QueryParameters.Select = ["c256x256"], cancellationToken);
```

However this generates a "wrong" URI for the thumbnail, which isn't suitable for getting the binary data AFAICS. It generates OData select, and it seemingly cannot be made to download content.

Both [get a single thumbnail](https://learn.microsoft.com/en-us/graph/api/driveitem-list-thumbnails?view=graph-rest-1.0&tabs=csharp#get-a-single-thumbnail) and [retrieve thumbnail binary content](https://learn.microsoft.com/en-us/graph/api/driveitem-list-thumbnails?view=graph-rest-1.0&tabs=csharp#retrieve-thumbnail-binary-content) have no documentation that pertains to the C# SDK. It says:

> Snippet not available

**Comment by andrueastman** [comment, marked as answer]
Thanks for raising this @Ghostbird 

This is a issue with the metadata used to generate the SDK. In the meantime you can work around this with a code sample like below.

```cs
var thumbnailContent = graphClient.Drives["drive-id"].Items["item-id"].Content.WithUrl("https://graph.microsoft.com/me/drive/items/{item-id}/thumbnails/{thumb-id}/{size}/content").GetAsync();
```

**Reply by Ghostbird** [reply]
Ah, brilliant! I'd tried `WithUrl` on several properties of the thumbnail API but the return type was never right. But of course, on the `ContentRequestBuilder` it has the right type!

**Reply by Ghostbird** [reply]
Note: The URI template in the answer omits the version[^1]. The URI must always include a _version_ part. The URI use above will throw `ODataError: Invalid version me`.

The corrected version of the URI template above would be: `https://graph.microsoft.com/v1.0/me/drive/items/{item-id}/thumbnails/{thumb-id}/{size}/content`.

In my case our software is not operating on behalf of a user, and I therefore use: `https://graph.microsoft.com/v1.0/drives/{drive-id}/items/{item-id}/thumbnails/{thumb-id}/{size}/content`

And because in my case I wanted to generate a custom template size, the final URI template I used is:
`https://graph.microsoft.com/v1.0/drives/{drive-id}/items/{item-id}/thumbnails/0/c256x256/content`

[^1]: Oddly in most Microsoft Graph API documentation examples, _including_ some on the documentation page about versioning, the version is omitted, and when you try to apply the examples they will not work.
